spring-cloud-azure-feature-management/src/main/java/com/microsoft/azure/spring/cloud/feature/manager/entities/FeatureSet.java

```diff
@@ -43,20 +50,46 @@ public FeatureSet() {
     public void setFeatureManagement(HashMap<String, Object> featureManagement) {
         setFeatures(featureManagement);
     }
-    
+
     public void addFeature(Feature feature) {
-        featureManagement.put(feature.getId(), feature);
+        featureManagement.put(feature.getKey(), feature);
     }
-    
+
     public void setFeatures(HashMap<String, Object> features) {
         if (features == null) {
             return;
         }
-        for (String key: features.keySet()) {
-            if (features.get(key) instanceof Boolean) {
-                onOff.put(key, (Boolean) features.get(key));
+        for (String key : features.keySet()) {
+            addToFeatures(features, key, "");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void addToFeatures(HashMap<String, Object> features, String key, String combined) {
+        Object featureKey = features.get(key);
+        if (!combined.isEmpty() && !combined.endsWith(".")) {
+            combined += ".";
+        }
+        if (featureKey instanceof Boolean) {
+            onOff.put(combined + key, (Boolean) featureKey);
+        } else {
+            Feature feature = null;
+            try {
+                feature = mapper.convertValue(featureKey, Feature.class);
+            } catch (IllegalArgumentException e) {
+                logger.error("Found invalid feature {} with value {}.", combined + key, featureKey.toString());
+            }
+
+            // When coming from a file "feature.flag" is not a possible flag name
+            if (feature != null && feature.getEnabledFor() == null && feature.getKey() == null) {
+                if (LinkedHashMap.class.isAssignableFrom(featureKey.getClass())) {
+                    features = (LinkedHashMap<String, Object>) featureKey;
+                    for (Object fKey : features.keySet().toArray()) {
```

Comment:
What is the purpose of calling toArray()? Why can't we just iterate features.keyset()